### DISCUSSION TITLE
Should leader automatically commit a uniform membership if the last config is a joint config?

### DISCUSSION THREAD

**Discussion by drmingdrmer** [databendlabs/openraft#17] [Ideas]
If a 2-phase membership change did not finish, i.e., a joint config is left in raft log.
Should a new leader detect this and commit a uniform membership config?

- Pros: Auto commit uniform config will leave a clean env for user. This is what the etcd-raft does.

- Cons: joint config is not actually special at all. Not treat it as a special case just generalize raft algo: it can be used to setup a two-DC cluster: `[{a,b,c}, {x,y,z}]` and a quorum is union of all majority in every DC, e.g. `[{a,b}, {y,z}]` is a quorum. Further more, in a three-DC cluster, it impl a [hierarchical quorum](https://zookeeper.apache.org/doc/r3.5.9/zookeeperHierarchicalQuorums.html), just like zookeeper does.

**Comment by jkosh44** [comment]
I know this is a fairly old post, but maybe would could add a config that toggles this behavior on or off?

**Reply by drmingdrmer** [reply]
@jkosh44 Thanks for picking this up. I'd rather not add this to the core, even behind a toggle: openraft treats a joint config as a legitimate final state, not an intermediate one that needs cleanup. A cluster is allowed to run in a joint config indefinitely, for example to build the multi-DC quorums described in the original post.

An application that wants a clean single-config setup can do this itself: read the current membership config from the Raft metrics, and propose a new uniform config if a joint one is left over. A few application-level lines cover the whole task, so building the same logic into the core would only add complexity.
